drivers/text_display/text_display_grove_lcd.c

```diff
@@ -0,0 +1,355 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//#define DT_DRV_COMPAT seeed_grove_lcd_rgb
+#define DT_DRV_COMPAT grove_lcd
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/text_display.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/misc/grove_lcd/grove_lcd.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(text_display_grove_lcd, CONFIG_TEXT_DISPLAY_LOG_LEVEL);
+
+#define GROVE_RGB_BACKLIGHT_ADDR        (0x62)
+
+/* GLCD_CMD_SCREEN_CLEAR has no options */
+/* GLCD_CMD_CURSOR_RETURN has no options */
+
+/* Defines for the GLCD_CMD_CURSOR_SHIFT */
+#define GLCD_CS_DISPLAY_SHIFT		(1 << 3)
+#define GLCD_CS_RIGHT_SHIFT		(1 << 2)
+
+/* LCD Display Commands */
+#define GLCD_CMD_SCREEN_CLEAR		(1 << 0)
+#define GLCD_CMD_CURSOR_RETURN		(1 << 1)
+#define GLCD_CMD_INPUT_SET		(1 << 2)
+#define GLCD_CMD_DISPLAY_SWITCH		(1 << 3)
+#define GLCD_CMD_CURSOR_SHIFT		(1 << 4)
+#define GLCD_CMD_FUNCTION_SET		(1 << 5)
+#define GLCD_CMD_SET_CGRAM_ADDR		(1 << 6)
+#define GLCD_CMD_SET_DDRAM_ADDR		(1 << 7)
+#define GLCD_DS_DISPLAY_ON              (1 << 2)
+#define GLCD_DS_CURSOR_ON               (1 << 1)
+#define GLCD_DS_BLINK_ON                (1 << 0)
+
+#define REGISTER_R	0x04
+#define REGISTER_G	0x03
+#define REGISTER_B	0x02
+
+struct grove_lcd_data {
+	uint8_t input_set;
+	bool power;
+	bool cursor;
+	bool blinking;
+	uint8_t function;
+	uint8_t backlight;
+};
+
+struct grove_lcd_config {
+	struct i2c_dt_spec bus;
+};
+
+static uint8_t colour_define[][3] = {
+	{ 255, 255, 255 },	/* white */
+	{ 255, 0,   0   },      /* red */
+	{ 0,   255, 0   },      /* green */
+	{ 0,   0,   255 },      /* blue */
+};
+
+static void grove_lcd_rgb_reg_set(const struct device *i2c, uint8_t addr, uint8_t dta)
+{
+	uint8_t data[2] = { addr, dta };
+
+	i2c_write(i2c, data, sizeof(data), GROVE_RGB_BACKLIGHT_ADDR);
+}
+
+static int grove_lcd_print(const struct device *dev, const uint8_t *data, uint16_t size)
+{
+	int rc = 0;
+	const struct grove_lcd_config *config = dev->config;
+	uint8_t buf[] = { GLCD_CMD_SET_CGRAM_ADDR, 0 };
+	int16_t i;
+
+	for (i = 0; i < size; i++) {
+		buf[1] = data[i];
+		rc = i2c_write_dt(&config->bus, buf, sizeof(buf));
+	}
+
+	return rc;
+}
+
+static int grove_lcd_cursor_position_set(const struct device *dev, uint8_t type, uint8_t x, uint8_t y)
+{
+	const struct grove_lcd_config *config = dev->config;
+
+	unsigned char data[2];
+
+	if (y == 0U) {
+		x |= 0x80;
+	} else {
+		x |= 0xC0;
+	}
+
+	data[0] = GLCD_CMD_SET_DDRAM_ADDR;
+	data[1] = x;
+
+	return i2c_write_dt(&config->bus, data, 2);
+}
+
+static int grove_lcd_clear(const struct device *dev)
+{
+	int rc;
+	const struct grove_lcd_config *config = dev->config;
+	uint8_t clear[] = { 0, GLCD_CMD_SCREEN_CLEAR };
+
+	rc = i2c_write_dt(&config->bus, clear, sizeof(clear));
+	LOG_DBG("clear, delay 20 ms");
+	k_sleep(K_MSEC(20));
```

Comment:
This is from the original zephyr driver (copy and paste) and would think sleeping is the only course since the command should be blocking until it is complete (someone raised this in the architecture meeting)

Reply:
Commands can still be blocking if you use e.g. a semaphore (semaphore can be unlocked via ISR or workqueue).

This is done in a few places in e.g. sensor drivers.

Just a thought.